### PR TITLE
Added a listerner when new terms of service is created

### PR DESCRIPTION
**Pull request by SagarGi**
### Description
This PR adds events called `NewTermsCreatedEvent` and `SignaturesResetEvent` when a terms of service is created.

### Why this event needed?
We have a nextcloud application [integration_openproject](https://github.com/nextcloud/integration_openproject) which required this event to be created for one of its feature. 
The feature of this application has such a user "OpenProject" who does not login but does some api request through app password token provided by nextcloud. To sign the Terms of Services a user has to login and then sign those. We are partly able to sign the terms of services through code from the application itself. For that we have to cover three steps in order to sign the Terms of services for the user "OpenProject"

>TOS=Terms Of Services

1. Sign TOS for user "OpenProject" at the time when `terms_of_services` app is enabled
2. Sign TOS for user "OpenProject" at the time it is created and app is enabled already
3. Sign TOS for user "OpenProject" when new TOS is created.

We can cover and handle the first 2 above scenarios but we fail to sign the TOS for 3rd scenario when new TOS is created since there is no event that the app can listen into to sign those TOS for user "OpenProject".

This is just a request or proposal if this application could add some event listener.
Created a new event in this repository would be really appreciated.
@nickvergessen

### CODE REVIEW

File: lib/Events/NewTOSCreatedEvent.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+use OCP\EventDispatcher\Event;
+
+/**
+ * Class MoveToTrashEvent
+ *
+ * Event to allow other apps to disable the trash bin for specific files
+ *
+ * @package OCA\Terms_Of_Service\Events
+ */
+class NewTOSCreatedEvent extends Event {
```
Comment (nickvergessen) [inline lib/Events/NewTOSCreatedEvent.php:36]: Following the existing naming convention this should be:
```suggestion
class TermsCreatedEvent extends Event {
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]
I guess you would also need an event (`SignaturesResetEvent`) in 
https://github.com/nextcloud/terms_of_service/blob/d03629a059438194310f773762cfbc644a9f7eb8/lib/Controller/SigningController.php#L116
to cover all of your usecases.

**Comment by SagarGi** [comment]
> I guess you would also need an event (`SignaturesResetEvent`) in
> 
> https://github.com/nextcloud/terms_of_service/blob/d03629a059438194310f773762cfbc644a9f7eb8/lib/Controller/SigningController.php#L116
> 
> 
> to cover all of your usecases.

@nickvergessen oh yes thanks :smiley: . Resetting clears up the data for the signed users. Will be creating it right away.

File: lib/Events/ResetSignaturesEvent.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+use OCP\EventDispatcher\Event;
+
+/**
+ * Class MoveToTrashEvent
+ *
+ * Event to allow other apps to disable the trash bin for specific files
+ *
+ * @package OCA\Terms_Of_Service\Events
+ */
+class ResetSignaturesEvent extends Event {
```
Comment (nickvergessen) [inline lib/Events/ResetSignaturesEvent.php:36]:
```suggestion
class SignaturesResetEvent extends Event {
```

File: lib/Events/SignaturesResetEvent.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017 Bjoern Schiessle <bjoern@schiessle.org>
+ *
+ * @author Arthur Schiwon <blizzz@arthur-schiwon.de>
+ * @author Bjoern Schiessle <bjoern@schiessle.org>
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
```
Comment (nickvergessen) [inline lib/Events/SignaturesResetEvent.php:7]:
```suggestion
 * @copyright Copyright (c) 2024 Sagar Gurung  <sagargurung1001@gmail.com>
 *
 * @author Sagar Gurung  <sagargurung1001@gmail.com>
```

File: lib/Events/TermsCreatedEvent.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017 Bjoern Schiessle <bjoern@schiessle.org>
+ *
+ * @author Arthur Schiwon <blizzz@arthur-schiwon.de>
+ * @author Bjoern Schiessle <bjoern@schiessle.org>
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
```
Comment (nickvergessen) [inline lib/Events/TermsCreatedEvent.php:7]:
```suggestion
 * @copyright Copyright (c) 2024 Sagar Gurung  <sagargurung1001@gmail.com>
 *
 * @author Sagar Gurung  <sagargurung1001@gmail.com>
```

**Review by nickvergessen — APPROVED** [review]

**Comment by SagarGi** [comment]
@nickvergessen Also app [integration_openproject](https://github.com/nextcloud/integration_openproject) supports the latest 3 version of Nextcloud [25 (will drop soon) , 26 , 27, 28]. So this events needs to be in `terms_of_service` app for the supported NC versions if i am not wrong?.

**Comment by nickvergessen** [comment]
Should be yeah

**Comment by nickvergessen** [comment]
But I'm not maintaining the app anymore, so pinging @GretaD and @max-nextcloud

File: lib/Events/SignaturesResetEvent.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+use OCP\EventDispatcher\Event;
+
+/**
+ * Class MoveToTrashEvent
+ *
+ * Event to allow other apps to disable the trash bin for specific files
+ *
+ * @package OCA\Terms_Of_Service\Events
+ */
```
Comment (max-nextcloud) [inline lib/Events/SignaturesResetEvent.php:33]: docs are from a different event.

File: lib/Events/TermsCreatedEvent.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+use OCP\EventDispatcher\Event;
+
+/**
+ * Class MoveToTrashEvent
+ *
+ * Event to allow other apps to disable the trash bin for specific files
+ *
+ * @package OCA\Terms_Of_Service\Events
+ */
```
Comment (max-nextcloud) [inline lib/Events/TermsCreatedEvent.php:33]: docs are from a different event.

**Review by max-nextcloud — COMMENTED** [review]
All in all looks good to me. Some minor comments about docs.

Usually I am not fond of backporting enhancements to older versions - but okay to make an exception here for the versions still in maintenance (i.e. >= 26).

**Comment by SagarGi** [comment]
> All in all looks good to me. Some minor comments about docs.
> 
> Usually I am not fond of backporting enhancements to older versions - but okay to make an exception here for the versions still in maintenance (i.e. >= 26).

Thanks. we are dropping the support for NC-25 very soon. Its good to have these from NC version>=26.

**Review by max-nextcloud — APPROVED** [review]

**Pull request merged**